src/Microsoft.ML.KMeansClustering/KMeansPredictor.cs

```diff
@@ -275,5 +280,78 @@ public void GetClusterCentroids(ref VBuffer<Float>[] centroids, out int k)
         }
+
+        public bool SaveAsOnnx(OnnxContext ctx, string[] outputNames, string featureColumn)
+        {
+            // Computation graph of distances to all centriods for a batch of exmples. Note that a centriod is just
+            // the center of a cluster. We use [] to denote the dimension of a variable; for example, X [3, 2] means
+            // that X is a 3-by-2 tensor. In addition, for a matrix X, X^T denotes its transpose.
+            //
+            // Symbols:
+            // l: # of examples
+            // n: # of features per input example
+            // X: input examples, l-by-n tensor.
+            // C: centriods, k-by-n tensor.
+            // C^2: 2-norm of all centriod vectors, its shape is [k].
+            // Y: 2-norm of difference between examples and centriods, l-by-k tensor. The value at i-th row and k-th
+            // column row, Y[i,k], is the distance from example i to centrioid k.
+            // L: the id of the nearest centriod for each input example, its shape is [l].
+            //
+            // .------------------------------------------------------.
+            // |                                                      |
+            // |                                                      v
+            // X [l, n] --> ReduceSumSquare --> X^2 [l]             Gemm (alpha=-2, transB=1) <-- C [k, n]
+            //                                   |                    |
+            //                                   |                    v
+            //                                   `------> Add <---- -2XC^T [l, k]
+            //                                             |
+            //                                             v
+            //                                             Z [l, k] ----------> Add <------------C^2 [k]
+            //                                                                   |
+            //                                                                   v
+            //                                           L [l] <--- ArgMin <---  Y [l, k]
+
+            // Allocate C, which is a constant tensor in prediction phase
+            var shapeC = new long[] { _centroids.Length, _centroids[0].Length };
+            var tensorC = new List<float>();
+            foreach (var centriod in _centroids)
+                tensorC.AddRange(centriod.DenseValues());
+            var nameC = ctx.AddInitializer(tensorC, shapeC, "C");
+
+            // Save C^2 as an initializer because it's a constant.
+            var shapeC2 = new long[] { _centroidL2s.Length };
+            var nameC2 = ctx.AddInitializer(_centroidL2s, shapeC2, "C2");
+
+            // Retrieve the name of X
+            var nameX = featureColumn;
+
+            // Compute X^2 from X
+            var nameX2 = ctx.AddIntermediateVariable(null , "X2", true);
+            var reduceNodeX2 = ctx.CreateNode("ReduceSumSquare", nameX, nameX2, ctx.GetNodeName("ReduceSumSquare"), "");
```

Comment:
>"" [](start = 116, length = 2)

If CreateNode has a default domain value of a null string to represent the default domain, what is the point of an empty string being passed in for the domain? Is this meant to be a non default case? Or is the default parameter wrong?

Reply:
It's not a default case. Default domain is "ai.onnx.ml", not "" (which means "ai.onnx" in ONNX spec) we need here.